File: Algorithms/121. Best Time to Buy and Sell Stock.ts
function maxProfit(prices: number[]): number {
    if (prices.length < 2) {
        return 0;
    }

    // Dynamic Programming Array
    let dp: number[] = new Array(prices.length).fill(0);
    let minPrice = prices[0];

    for (let i = 1; i < prices.length; i++) {
        dp[i] = Math.max(dp[i - 1], prices[i] - minPrice);
        minPrice = Math.min(minPrice, prices[i]);
    }

    return dp[prices.length - 1];
};

File: Algorithms/121. Best Time to Buy and Sell Stock.dart
import 'dart:math';

class Solution {
  int maxProfit(List<int> prices) {
    if (prices.length < 2) {
      return 0;
    }

    // Dynamic Programming Array
    List<int> dp = List.filled(prices.length, 0);

    int minPrice = prices[0];

    for (int i = 1; i < prices.length; i++) {
      dp[i] = max(dp[i - 1], prices[i] - minPrice);
      minPrice = min(minPrice, prices[i]);
    }

    return dp[prices.length - 1];
  }
}
